### PR TITLE
feat: add Idempotency to Postgres

### LINKED ISSUE #7749: Add Idempotency support to Postgres
### New Feature / Enhancement Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Current Limitation

Postgres doesn't support Idempotency, mainly because it doesn't have a ttl index.

### Feature / Enhancement Description

Originally discussed in #6744

### Example Use Case

Filters out duplicate requests due to bad connectivity.

### Alternatives / Workarounds

Clone the repo and manually edit lines that don't create/manage the `_Idempotent` class on Postgres. Then, manually create trigger or cron job that deletes rows on the database.

### 3rd Party References

Stack discusses a way to [add triggers to Postgres](https://stackoverflow.com/questions/26046816/is-there-a-way-to-set-an-expiry-time-after-which-a-data-entry-is-automaticall) to achieve similar functionality to Mongo ttl index.

### PR DESCRIPTION
**Pull request by cbaker6**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description

Idempotency currently isn't enabled for Postgres. This is due to no ttl index support. Close #7749

Related issue: #7749 

### Approach

Add a Postgres function that can be called using cron, [pgAdmin](https://www.pgadmin.org/docs/pgadmin4/development/pgagent_jobs.html), or similar by calling the query `SELECT idempodency_delete_old_rows()`. Example script:

```bash
#!/bin/bash

set -e
psql -v ON_ERROR_STOP=1 --username "$POSTGRES_USER" --dbname "$POSTGRES_DB" <<-EOSQL
    SELECT idempodency_delete_old_rows();
EOSQL

exec "$@"
```

### TODOs before merging

- [x] Enable idempotency tests for Postgres
- [x] Add additional tests for idempotency for Postges
- [x] A changelog entry is created automatically using the pull request title (do not manually add a changelog entry)

### CODE REVIEW

**Comment by cbaker6** [comment]
@mtrezza this is ready for review. This requires an additional step to be taken by the developer when compared to the mongo implementation, but I tried to simplify the process. Let me know if you have questions.

File: README.md [hunk 1]
```diff
@@ -527,7 +527,25 @@ let api = new ParseServer({
 
 ### Notes <!-- omit in toc -->
 
-- This feature is currently only available for MongoDB and not for Postgres.
+- There are no additional steps needed to use this feature with MongoDB
+- For Postgres, you will need to create a cron job, use [pgAdmin](https://www.pgadmin.org/docs/pgadmin4/development/pgagent_jobs.html), or similar to call the Postgres function, `idempodency_delete_old_rows()`. For example, the cron job can call the script below (make sure the script has the same priveledges to log into Postgres as your Parse Server):
```
Comment (mtrezza) [inline README.md]: could you reword to:

> - For MongoDB, no additional steps are needed.
> - For Postgres, you need to create a cron job using [pgAdmin](https://www.pgadmin.org/docs/pgadmin4/development/pgagent_jobs.html) or similar to call a Postgres function `idempotency_delete_expired_records()` that deletes expired idempotency records. You can find an example script below. Make sure the script has the same privileges to log into Postgres as Parse Server.

i think "expired_records" is more universal than "old_rows". in case a future no-sql DB adapter is added and also doesn't support TTL, "row" may not be an accurate term and we'd end up with different functions names for different DB adapters.

File: README.md [hunk 2]
```diff
@@ -527,7 +527,25 @@ let api = new ParseServer({
...
+
+set -e
+psql -v ON_ERROR_STOP=1 --username "$POSTGRES_USER" --dbname "$POSTGRES_DB" <<-EOSQL
+    SELECT idempodency_delete_old_rows();
+EOSQL
+
+exec "$@"
+```
+
+Assuming the script above is named, `parse_idempodency_delete_old_rows.sh`, a cron job that runs the script every 2 minutes may look like:
```
Comment (mtrezza) [inline README.md]: could you rename to

> parse_idempotency_delete_expired_records.sh

File: README.md (hunk 1 above)
Comment (cbaker6) [reply README.md]: Updated

File: README.md (hunk 2 above)
Comment (cbaker6) [reply README.md]: Updated

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 3]
```diff
@@ -2416,25 +2407,39 @@ export class PostgresStorageAdapter implements StorageAdapter {
...
-        error.code === PostgresUniqueIndexViolationError &&
-        error.message.includes(indexNameOptions.name)
-      ) {
-        // Cast the error into the proper parse error
-        throw new Parse.Error(
-          Parse.Error.DUPLICATE_VALUE,
-          'A duplicate value for a field with unique values was provided'
-        );
-      } else {
+    const setIdempodencyFunction = options.setIdempodencyFunction !== undefined ? options.setIdempodencyFunction : false;
```
Comment (mtrezza) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Could you just fix the typo and replace all "idempo`d`ency" with "idempo`t`ency"?

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js (hunk 3 above)
Comment (cbaker6) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: done

**Review by mtrezza — APPROVED** [review]
~~Looks good!~~

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 4]
```diff
@@ -15,7 +15,6 @@ const PostgresRelationDoesNotExistError = '42P01';
 const PostgresDuplicateRelationError = '42P07';
 const PostgresDuplicateColumnError = '42701';
 const PostgresMissingColumnError = '42703';
-const PostgresDuplicateObjectError = '42710';
```
Comment (mtrezza) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: From your issue description it seems that this PR contains unrelated changes. Could you please move these into a separate PR?

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js (hunk 4 above)
Comment (cbaker6) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: done

File: src/Controllers/DatabaseController.js [hunk 5]
```diff
@@ -1751,18 +1752,28 @@ class DatabaseController {
...
+        })
         .catch(error => {
-          logger.warn('Unable to ensure uniqueness for idempotency request ID: ', error);
+          logger.warn('Unable to create TTL index for idempotency expire date: ', error);
           throw error;
         });
-
+    } else if (this.adapter instanceof PostgresStorageAdapter) {
+      const options = this.idempotencyOptions;
+      options.setIdempotencyFunction = true;
```
Comment (mtrezza) [inline src/Controllers/DatabaseController.js]: If I see the diff correctly, the code block for postgres seems to be almost a duplicate of the one for mongo, expect for `options.setIdempotencyFunction = true;`. Could you dedupe this?

Comment (cbaker6) [reply src/Controllers/DatabaseController.js]: I tried before, but the mongo adapter began failing for some reason I couldn't figure out. I'll try again to see if the problem I had persists, or if I can implement it a different way

Comment (cbaker6) [reply src/Controllers/DatabaseController.js]: Lets see if this change passes on mongo and let me know if it looks good

Comment (cbaker6) [reply src/Controllers/DatabaseController.js]: Looks like it's good to go...

**Comment by mtrezza** [comment]
Could you just resolve the conflicts and then lets see if the tests pass.

**Review by mtrezza — COMMENTED** [review]
Could you just resolve the conflicts and then let's see if the tests pass.

**Review by mtrezza — APPROVED** [review]

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.0.0-alpha.16](https://github.com/parse-community/parse-server/releases/tag/5.0.0-alpha.16)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.0.0-beta.10](https://github.com/parse-community/parse-server/releases/tag/5.0.0-beta.10)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.1.0](https://github.com/parse-community/parse-server/releases/tag/5.1.0)
